### PR TITLE
udevadm: also show write-only attributes

### LINKED ISSUE #10102: [udev] libsystemd / libudev: no way to find out about write-only attributes
**systemd version the issue has been seen with**
232-25+deb9u4 on Debian, eudev as used in OpenWRT, but present in v239.

**Used distribution**
 Debian, OpenWRT

**Expected behaviour you didn't see**
I expected to be able to use libudev to see write-only attributes for a device.

I ran into this while working with a PWM (trying to use /sys/class/pwm/pwmchip0/export). Unfortunately, this takes some specific hardware to reproduce, but a similar situation is seen in the GPIO class (/sys/class/gpio/export)

Another example is /sys/devices/system/memory/soft_offline_page

**Unexpected behaviour you saw**
I get all of the attributes I expect, except for the ones that I don't have read permission for. This specifically hit me in libudevpp, which was getting its list of attributes from libudev, which was getting its list from libsystemd.

**Steps to reproduce the problem**
Use device_sysattrs_read_all on a device with a write-only attribute.
I did a quick test with udevadm, but I had to bypass a couple of checks to get it to print attributes with write-only values. On my development machine, I can look at /sys/class/gpio, and I don't get "export" back as an attribute.

I believe the issue is stemming from a specific check that requires read permission in [sd-device.c](https://github.com/systemd/systemd/blame/32397af3d435aab0895889d9b89a6950b6246a80/src/libsystemd/sd-device/sd-device.c#L1613). Taking that out even allows me to get a modified udevadm to see the attributes.

I'm not saying that's a terrible check to have in there, but unless there's some other way to see these attributes, that leaves us no way to ever know we can write to them.

So I suppose the questions are:
* Can we take this check out? Will that break a hundred different things that rely on the current functionality?
* If we can't change the current behavior, should we add some other way to access this write-only attributes? Or is this out of the scope of libudev?

(Interestingly, udevadm actually handles the case of NULL values just fine, even though it was never possible with that check in place.)

### PR DESCRIPTION
**Pull request by yuwata**
Closes #10102.

### CODE REVIEW

**Comment by keszybz** [comment]
I don't think we need a separate interface for this. It feels very ad-hoc. We already have  `sd_device_get_sysattr_value()`, and the caller needs to be prepared to get an error, including EPERM or ENOENT if the device was already removed. I think we should just can just relax the check for readability, and return also the unreadable attributes in `sd_device_get_sysattr_{first,next}()`.

It would also be nice to document those functions…

**Comment by yuwata** [comment]
OK. Now both readable and write-only attributes are listed by `FOREACH_DEVICE_SYSATTR()`. PTAL.

**Comment by keszybz** [comment]
LGTM.

I'll wait a bit for comments, but I think this can be merged.

**Comment by yuwata** [comment]
OK. Then, let's set the green label and wait for other reviewers to review and merge this.

File: src/libsystemd/sd-device/sd-device.c [hunk 1]
```diff
@@ -1812,7 +1812,7 @@ static int device_sysattrs_read_all_internal(sd_device *device, const char *subd
                 if (lstat(path, &statbuf) != 0)
                         continue;
 
-                if (!(statbuf.st_mode & S_IRUSR))
+                if ((statbuf.st_mode & (S_IRUSR | S_IWUSR)) == 0)
```
Comment (medhefgo) [inline src/libsystemd/sd-device/sd-device.c:1815]: `FLAGS_SET`?

Comment (yuwata) [reply src/libsystemd/sd-device/sd-device.c:1815]: No. `FLAGS_SET()` returns true only when all flags are set. So,
`FLAGS_SET(statbuf.st_mode, S_IRUSR | S_IWUSR)` is true only when the file is rw.

File: src/udev/udevadm-info.c [hunk 2]
```diff
@@ -96,18 +97,22 @@ static int print_all_attributes(sd_device *device, bool is_parent) {
...
+                        if (len > 0)
+                                continue;
 
-                /* skip nonprintable attributes */
-                len = strlen(value);
-                while (len > 0 && isprint((unsigned char) value[len-1]))
-                        len--;
-                if (len > 0)
+                } else if (r == -EPERM)
+                        value = "(write-only)";
```
Comment (poettering) [inline src/udev/udevadm-info.c:114]: that's a stretch to claim. We don't know if the file is wrtiable. All we know is that it is not readable. Might be totally inaccessible due to some MAC (bpf-lsm). I'd change this to "(not readable)"

Comment (poettering) [reply src/udev/udevadm-info.c:114]: #22372

**Pull request merged**